frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseRepositoryAction.java

```diff
@@ -0,0 +1,292 @@
+
+  @Untainted
+  @I18nFormat({I18nConversionCategory.GENERAL})
+  final String actionIdFormatString = getNonHtmlString("action.GitMachete.TraverseRepositoryAction.invoked-action.id");
+
+  @Override
+  public LambdaLogger log() {
+    return LOG;
+  }
+
+  public static final String SHOW_PULL_APPROVAL = "git-machete.pull.approval.show";
+  public static final String SHOW_PUSH_APPROVAL = "git-machete.push.approval.show";
+  public static final String SHOW_TRAVERSE_APPROVAL = "git-machete.traverse.approval.show";
+
+  @Override
+  @UIEffect
+  protected void onUpdate(AnActionEvent anActionEvent) {
+    super.onUpdate(anActionEvent);
+    val presentation = anActionEvent.getPresentation();
+    if (!presentation.isEnabledAndVisible()) {
+      return;
+    }
+    presentation.setDescription(getNonHtmlString("action.GitMachete.TraverseRepositoryAction.description.text"));
+
+    val repositorySnapshot = getGitMacheteRepositorySnapshot(anActionEvent);
+    val branchLayout = repositorySnapshot != null ? repositorySnapshot.getBranchLayout() : null;
+
+    if (branchLayout == null || branchLayout.getRootEntries().isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription(getNonHtmlString("action.GitMachete.TraverseRepositoryAction.description.empty-layout"));
+      return;
+    }
+  }
+
+  @Override
+  @UIEffect
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    val repository = getSelectedGitRepository(anActionEvent);
+    val branchLayout = getBranchLayout(anActionEvent);
+
+    if (branchLayout != null && branchLayout.getRootEntries().nonEmpty() && repository != null) {
+      boolean traverseApproved = true;
+      if (PropertiesComponent.getInstance().getBoolean(SHOW_TRAVERSE_APPROVAL, /* defaultValue */ true)) {
+
+        val traverseApprovalDialogBuilder = MessageDialogBuilder.okCancel(
+            getString("action.GitMachete.TraverseRepositoryAction.dialog.traverse-approval.title"),
+            getString("action.GitMachete.TraverseRepositoryAction.dialog.traverse-approval.text.HTML"));
+
+        traverseApprovalDialogBuilder
+            .yesText(getString("action.GitMachete.TraverseRepositoryAction.dialog.traverse-approval.ok-text"))
+            .icon(Messages.getQuestionIcon()).doNotAsk(new TraverseApprovalDialog());
+
+        traverseApproved = traverseApprovalDialogBuilder.ask(getProject(anActionEvent));
+      }
+
+      if (traverseApproved) {
+        var currentBranchName = getCurrentBranchNameIfManaged(anActionEvent);
+        var previousEntry = currentBranchName != null ? branchLayout.findPreviousEntry(currentBranchName) : null;
+        while (previousEntry != null) {
+          val checkoutPreviousAction = ActionManager.getInstance()
+              .getAction(actionIdFormatString.format(CheckoutPreviousAction.class.getSimpleName()));
+          checkoutPreviousAction.actionPerformed(anActionEvent);
```

Comment:
Checkouts in a loop? ➿ that might take very long!

Reply:
I'd rather just check out the first branch to be worked upon

Reply:
yes, gonna change that to checking out the first branch in background able chained by other actions onSuccess